File: 2014년문제답안/Problem02/Graduates/par2.f
c234567
      function potential(r, np, jstart, jfinal)
      double precision potential
      real r(3,np)
      integer np, jstart, jfinal
      potential = 0
      do j=jstart+1, jfinal
         do i = j+1, np
            x = r(1,i) - r(1,j)
            y = r(2,i) - r(2,j)
            z = r(3,i) - r(3,j)
            dist = sqrt(x*x+y*y+z*z)
            if(dist .gt.0) potential = potential + 1.d0/dist
         enddo
      enddo
      jjstart = min(np-jfinal,np/2)
      jjfinal = min(np/2, np-jstart)
      do j = jjstart + 1, jjfinal 
         do i = j +1, np
            x = r(1,i) - r(1,j)
            y = r(2,i) - r(2,j)
            z = r(3,i) - r(3,j)
            dist = sqrt(x*x+y*y+z*z)
            if(dist .gt.0) potential = potential + 1.d0/dist
         enddo
      enddo
      return
      end

      function getrandomnp(istep, niter)
      integer getrandomnp
      integer iseed
      common /seed/ iseed
      if(mod(istep-1,3).eq.0) then
         getrandomnp = (100000*ran2(iseed))
      else
         getrandomnp = (500*ran2(iseed))
      endif
      return
      end
      program mpi_c
      include 'mpif.h'
      parameter(maxnp=5000000)
	  integer iREADY,iWRITING,iFINISH,NP_TAG,iR_TAG, iPOT_TAG
      parameter(iREADY=0,iWRITING=-1,iFINISH=-991,NP_TAG=99,iR_TAG=990)
      parameter(iPOT_TAG=123)
      integer i,j,k
      integer np,niter
      real r(3,maxnp)
      double precision totpotent,potent,potential
      integer finish,ready
      integer iseed
      common /seed/ iseed
      real timearray(2), time1,time2,ran2
	  external ran2,potential
	  integer*8 nlocalsize, nwork, nsplit, njump

      integer mstatus(MPI_STATUS_SIZE), request,ierror
	  external real etime


      finish = iFINISH

      call MPI_INIT(ierror)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, nid, ierror)
      call MPI_COMM_RANK(MPI_COMM_WORLD, myid, ierror)

      niter = 50*3
      iseed = -9

      xx = ran2(iseed)

      time1 = etime(timearray)
      nlocalsize = 1000*1000

      if(myid.eq.0) then
         nsend = 0
		 nrecv = 0
         do i = 1, niter
            np = getrandomnp(i,niter)
            do j = 1,  np
               r(1,j) = 2*ran2(iseed) -1
               r(2,j) = 2*ran2(iseed) -1
               r(3,j) = 2*ran2(iseed) -1
            enddo
            nwork = np*(np-1d0)/2d0
            nsplit = (nwork-1)/nlocalsize + 1
            if(nsplit .le. 1) then
               njump = np/2+1
            else
               njump = (np/2-1)/nsplit + 1
            endif
            do j = np, np/2, -njump
               jstart = max(np/2,j-njump)
               jfinal = j
20             continue
                  call MPI_PROBE(MPI_ANY_SOURCE, iREADY, MPI_COMM_WORLD,
     &                       mstatus, ierror)
                  call MPI_RECV(ready, 1, MPI_INTEGER, 
     &                       mstatus(MPI_SOURCE),iREADY,MPI_COMM_WORLD,
     &                       mstatus, ierror)
                  if(ready .eq. iREADY) then
                     nsend = nsend + 1
                     call MPI_SEND(np,1,MPI_INTEGER, mstatus(MPI_SOURCE)
     &                    , NP_TAG,MPI_COMM_WORLD,IERROR)
                     call MPI_SEND(r,np*3,MPI_REAL, mstatus(MPI_SOURCE),
     &                    iR_TAG,MPI_COMM_WORLD,IERROR)
                     call MPI_SEND(jstart,1,MPI_INTEGER, 
     &                    mstatus(MPI_SOURCE), iR_TAG,
     &                    MPI_COMM_WORLD,IERROR)
                     call MPI_SEND(jfinal,1,MPI_INTEGER, 
     &                 mstatus(MPI_SOURCE),iR_TAG,MPI_COMM_WORLD,IERROR)
                  else
                     nrecv = nrecv + 1
                     call MPI_RECV(potent, 1, MPI_DOUBLE_PRECISION, 
     &                 mstatus(MPI_SOURCE), iPOT_TAG,MPI_COMM_WORLD,
     &                 mstatus, IERROR)
                     totpotent = totpotent + potent
                  endif
               if(ready .ne. iREADY) goto 20
            enddo
         enddo
         j = 0
         i = nrecv
         do while(i .lt. nsend)
            call MPI_PROBE(MPI_ANY_SOURCE, iREADY, MPI_COMM_WORLD, 
     &                     mstatus, IERROR)
            call MPI_RECV(ready, 1, MPI_INTEGER, mstatus(MPI_SOURCE), 
     &                     iREADY, MPI_COMM_WORLD, mstatus, IERROR)
            if(ready .eq. iWRITING) then
               call MPI_RECV(potent, 1, MPI_DOUBLE_PRECISION, 
     &                  mstatus(MPI_SOURCE), iPOT_TAG, MPI_COMM_WORLD,
     &                    mstatus, IERROR)
               totpotent = totpotent + potent
               call MPI_SEND(finish, 1, MPI_INTEGER, 
     &              mstatus(MPI_SOURCE), NP_TAG, MPI_COMM_WORLD, IERROR)
               i = i + 1
            else
               call MPI_SEND(finish, 1, MPI_INTEGER, mstatus(MPI_SOURCE)
     &               , NP_TAG, MPI_COMM_WORLD, IERROR)
               j = j + 1
            endif
         enddo
         do i = 1, nid-j-1
            call MPI_RECV(ready, 1, MPI_INTEGER, MPI_ANY_SOURCE, iREADY,
     &               MPI_COMM_WORLD, mstatus, IERROR)
            call MPI_SEND(finish, 1, MPI_INTEGER, mstatus(MPI_SOURCE), 
     &           NP_TAG, MPI_COMM_WORLD, IERROR)
         enddo
      else
         ready = iREADY
         call MPI_SEND(ready, 1, MPI_INTEGER, 0, iREADY, MPI_COMM_WORLD,
     &         IERROR)
         call MPI_RECV(np, 1, MPI_INTEGER, 0, NP_TAG, MPI_COMM_WORLD, 
     &         mstatus, IERROR)
         do while (np .ne. iFINISH)
            call MPI_RECV(r, np*3, MPI_REAL, 0, iR_TAG, MPI_COMM_WORLD, 
     &         mstatus, IERROR)
            call MPI_RECV(jstart, 1, MPI_INTEGER, 0, iR_TAG, 
     &           MPI_COMM_WORLD, mstatus, IERROR)
            call MPI_RECV(jfinal, 1, MPI_INTEGER, 0, iR_TAG, 
     &               MPI_COMM_WORLD, mstatus, IERROR)
            potent = potential(r, np, jstart, jfinal)
            ready = iWRITING
            call MPI_ISSEND(ready,1,MPI_INTEGER,0,iREADY,MPI_COMM_WORLD,
     &                    request, IERROR)
            call MPI_WAIT(request, mstatus,IERROR)
            call MPI_SEND(potent, 1, MPI_DOUBLE_PRECISION, 0, iPOT_TAG, 
     &              MPI_COMM_WORLD, IERROR)

            ready = iREADY
            call MPI_SEND(ready, 1, MPI_INTEGER, 0, iREADY, 
     &            MPI_COMM_WORLD,IERROR)
            call MPI_RECV(np, 1, MPI_INTEGER, 0, NP_TAG, MPI_COMM_WORLD,
     &            mstatus, IERROR)
         enddo
      endif

      time2 = etime(timearray)
      if(myid.eq.0) print *, '[F] np=',nid,'TOP is', totpotent, 'in ', 
     &            (time2-time1),' seconds'
      call MPI_FINALIZE(IERROR)

	  stop
	  end

      FUNCTION ran2(idum)
      INTEGER idum,IM1,IM2,IMM1,IA1,IA2,IQ1,IQ2,IR1,IR2,NTAB,NDIV
      REAL ran2,AM,EPS,RNMX
      PARAMETER (IM1=2147483563,IM2=2147483399,AM=1./IM1,IMM1=IM1-1,
     *IA1=40014,IA2=40692,IQ1=53668,IQ2=52774,IR1=12211,IR2=3791,
     *NTAB=32,NDIV=1+IMM1/NTAB,EPS=1.2e-7,RNMX=1.-EPS)
      INTEGER idum2,j,k,iv(NTAB),iy
      SAVE iv,iy,idum2
      DATA idum2/123456789/, iv/NTAB*0/, iy/0/
      if (idum.le.0) then
        idum=max(-idum,1)
        idum2=idum
        do 11 j=NTAB+8,1,-1
          k=idum/IQ1
          idum=IA1*(idum-k*IQ1)-k*IR1
          if (idum.lt.0) idum=idum+IM1
          if (j.le.NTAB) iv(j)=idum
11      continue
        iy=iv(1)
      endif
      k=idum/IQ1
      idum=IA1*(idum-k*IQ1)-k*IR1
      if (idum.lt.0) idum=idum+IM1
      k=idum2/IQ2
      idum2=IA2*(idum2-k*IQ2)-k*IR2
      if (idum2.lt.0) idum2=idum2+IM2
      j=1+iy/NDIV
      iy=iv(j)-idum2
      iv(j)=idum
      if(iy.lt.1)iy=iy+IMM1
      ran2=min(AM*iy,RNMX)
      return
      END
C  (C) Copr. 1986-92 Numerical Recipes Software +)-*1a311.

File: 2014년문제답안/Problem02/Graduates/par2.c
#include<stdio.h>
#include<stdlib.h>
#include<stddef.h>
#include<string.h>
#include<math.h>
#include<sys/time.h>
#include<mpi.h>


#define READY 0
#define WRITING -1
#define FINISH -991
#define NP_TAG 99
#define R_TAG 990
#define POT_TAG 123


#define max(a,b) (a)>(b)? (a):(b)
#define min(a,b) (a)<(b)? (a):(b)

long iseed = -9;
float ran2(long *);

typedef struct Pos{
	float x,y,z;
}Pos;

double potential(Pos *r, int np, int jstart, int jfinal){
	int i,j;
	double potent = 0;
	for(j=jstart;j<jfinal;j++){
		for(i=j+1;i<np;i++){
			float x= r[i].x-r[j].x;
			float y= r[i].y-r[j].y;
			float z= r[i].z-r[j].z;
			float dist = sqrtf(x*x+y*y+z*z);
			if(dist > 0.) potent += 1.L/dist;
		}
	}
	int jjstart = min(np-jfinal,np/2);
	int jjfinal = min(np/2,np - jstart);
	for(j=jjstart;j<jjfinal;j++){
		for(i=j+1;i<np;i++){
			float x= r[i].x-r[j].x;
			float y= r[i].y-r[j].y;
			float z= r[i].z-r[j].z;
			float dist = sqrtf(x*x+y*y+z*z);
			if(dist > 0.) potent += 1./dist;
		}
	}
	return potent;
}

int getrandomnp(int istep, int niter){
	if(istep %3 ==0) {
		return (int)(100000*(ran2(&iseed)));
	}
	else {
		return (int)(500*(ran2(&iseed)));
	}
}

struct timeval tv;
float gettime(){
	static int startflag = 1;
	static double tsecs0, tsecs1;
	if(startflag) {
		(void ) gettimeofday(&tv, NULL);
		tsecs0 = tv.tv_sec + tv.tv_usec*1.0E-6;
		startflag = 0;
	}
	(void) gettimeofday(&tv, NULL);
	tsecs1 = tv.tv_sec + tv.tv_usec*1.0e-6;
	return (float) (tsecs1 - tsecs0);
}


int myid,nid;

int main(int argc, char **argv){
	int i, j;
	int np,niter,maxnp=5000000;
	Pos *r;
	double totpotent=0;
	double potent;
	int finish = FINISH;
	MPI_Status mstatus;
	MPI_Request request;


	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &nid);
	MPI_Comm_rank(MPI_COMM_WORLD, &myid);

	niter = 50*3;
	r = (Pos*)malloc(sizeof(Pos)*maxnp);
	ran2(&iseed);

	float time1, time2;
	time1 = gettime();

	long nlocalsize = 10000*1000;
	int jstart, jfinal;


	if(myid==0){
		int nsend,nrecv;
		int ready,src,dest;
		nsend = nrecv = 0;
		for(i=0;i<niter;i++){
			np =  getrandomnp(i,niter);
			for(j=0;j<np;j++){
				r[j].x = 2.*(ran2(&iseed))-1.;
				r[j].y = 2.*(ran2(&iseed))-1.;
				r[j].z = 2.*(ran2(&iseed))-1.;
			}
			long nwork = (long)np*((long)np-1)/2;
			long nsplit,njump;
			nsplit = (nwork-1)/nlocalsize + 1;
			if(nsplit <=1) {
				njump = np/2+1;
			}
			else {
				njump = (np/2-1)/nsplit+1;
			}
			/*
			if(np == 90583) {
				int kkkk = 1;
				while(kkkk) {
					kkkk = 1;
				}
			}
			*/
			for(j=np;j>np/2;j-=njump)
			{
				jstart = max(np/2,j-njump);
				jfinal = j;
				do {
					MPI_Probe(MPI_ANY_SOURCE, READY,MPI_COMM_WORLD, &mstatus);
					MPI_Recv(&ready,1, MPI_INT, mstatus.MPI_SOURCE, READY,MPI_COMM_WORLD,&mstatus);
					if(ready == READY){
						nsend ++;
						MPI_Send(&np,1, MPI_INT,mstatus.MPI_SOURCE, NP_TAG,MPI_COMM_WORLD);
						MPI_Send(r,np*sizeof(Pos), MPI_BYTE,mstatus.MPI_SOURCE, R_TAG,MPI_COMM_WORLD);
						MPI_Send(&jstart,sizeof(int), MPI_INT,mstatus.MPI_SOURCE, R_TAG,MPI_COMM_WORLD);
						MPI_Send(&jfinal,sizeof(int), MPI_INT,mstatus.MPI_SOURCE, R_TAG,MPI_COMM_WORLD);
/*
						printf("P%d is sending %d data to pid= %d with np= %d :    [%d , %d) && [%d , %d)\n",myid,i,mstatus.MPI_SOURCE,
								np,jstart,jfinal, min(np/2,np-jfinal), min(np/2,np-jstart));
*/
					}
					else {
						nrecv ++;
						MPI_Recv(&potent,1,MPI_DOUBLE,mstatus.MPI_SOURCE, POT_TAG,MPI_COMM_WORLD,&mstatus);
						totpotent += potent;
					}
				}while(ready != READY);
			}
		}
		j = 0;
		for(i=nrecv;i<nsend;){
			MPI_Probe(MPI_ANY_SOURCE, READY, MPI_COMM_WORLD, &mstatus);
			MPI_Recv(&ready, 1, MPI_INT, mstatus.MPI_SOURCE, READY, MPI_COMM_WORLD, &mstatus);
			if(ready == WRITING){
				MPI_Recv(&potent,1,MPI_DOUBLE,mstatus.MPI_SOURCE, POT_TAG,MPI_COMM_WORLD, &mstatus);
				totpotent += potent;
				MPI_Send(&finish,1, MPI_INT, mstatus.MPI_SOURCE, NP_TAG, MPI_COMM_WORLD);
				i++;
			}
			else {
				MPI_Send(&finish,1, MPI_INT, mstatus.MPI_SOURCE, NP_TAG, MPI_COMM_WORLD);
				j ++;
			}
		}
		for(i=1;i<nid-j;i++){
			MPI_Recv(&ready, 1, MPI_INT, MPI_ANY_SOURCE, READY, MPI_COMM_WORLD, &mstatus);
			MPI_Send(&finish, 1, MPI_INT, mstatus.MPI_SOURCE, NP_TAG, MPI_COMM_WORLD);
		}
	}
	else {
		int ready = READY;
		MPI_Send(&ready,1, MPI_INT, 0, READY, MPI_COMM_WORLD);
		MPI_Recv(&np, 1, MPI_INT, 0, NP_TAG, MPI_COMM_WORLD, &mstatus);
		while(np != FINISH){
			MPI_Recv(r, sizeof(Pos)*np, MPI_BYTE, 0, R_TAG, MPI_COMM_WORLD, &mstatus);
			MPI_Recv(&jstart, sizeof(int), MPI_INT, 0, R_TAG, MPI_COMM_WORLD, &mstatus);
			MPI_Recv(&jfinal, sizeof(int), MPI_INT, 0, R_TAG, MPI_COMM_WORLD, &mstatus);
			potent = potential(r, np,jstart,jfinal);
			ready = WRITING;
			MPI_Issend(&ready, 1, MPI_INT, 0, READY, MPI_COMM_WORLD, &request);
			MPI_Wait(&request, &mstatus);
			MPI_Send(&potent, 1, MPI_DOUBLE, 0, POT_TAG, MPI_COMM_WORLD);

			ready = READY;
			MPI_Send(&ready, 1, MPI_INT,0,READY, MPI_COMM_WORLD);
			MPI_Recv(&np, 1, MPI_INT, 0, NP_TAG, MPI_COMM_WORLD, &mstatus);
		}
	}
	time2 = gettime();
	if(myid==0) printf("[C] np=%d TPOT=%20.10f in Wtimes = %g (sec)\n",nid,totpotent, (time2-time1));
	MPI_Finalize();
}
#define IM1 2147483563
#define IM2 2147483399
#define AM (1.0/IM1)
#define IMM1 (IM1-1)
#define IA1 40014
#define IA2 40692
#define IQ1 53668
#define IQ2 52774
#define IR1 12211
#define IR2 3791
#define NTAB 32
#define NDIV (1+IMM1/NTAB)
#define EPS 1.2e-7
#define RNMX (1.0-EPS)

float ran2(long *idum)
{
	int j;
	long k;
	static long idum2=123456789;
	static long iy=0;
	static long iv[NTAB];
	float temp;

	if (*idum <= 0) {
		if (-(*idum) < 1) *idum=1;
		else *idum = -(*idum);
		idum2=(*idum);
		for (j=NTAB+7;j>=0;j--) {
			k=(*idum)/IQ1;
			*idum=IA1*(*idum-k*IQ1)-k*IR1;
			if (*idum < 0) *idum += IM1;
			if (j < NTAB) iv[j] = *idum;
		}
		iy=iv[0];
	}
	k=(*idum)/IQ1;
	*idum=IA1*(*idum-k*IQ1)-k*IR1;
	if (*idum < 0) *idum += IM1;
	k=idum2/IQ2;
	idum2=IA2*(idum2-k*IQ2)-k*IR2;
	if (idum2 < 0) idum2 += IM2;
	j=iy/NDIV;
	iy=iv[j]-idum2;
	iv[j] = *idum;
	if (iy < 1) iy += IMM1;
	if ((temp=AM*iy) > RNMX) return RNMX;
	else return temp;
}
#undef IM1
#undef IM2
#undef AM
#undef IMM1
#undef IA1
#undef IA2
#undef IQ1
#undef IQ2
#undef IR1
#undef IR2
#undef NTAB
#undef NDIV
#undef EPS
#undef RNMX
/* (C) Copr. 1986-92 Numerical Recipes Software 71.+I0>+. */
